File: tools/IconTool.Core/IcoHelper.cs
namespace IconTool.Core;

/// <summary>
/// ICO 文件格式解析与操作工具。
/// </summary>
public static class IcoHelper
{
    /// <summary>
    /// 解析 ICO 文件字节，返回所有图像条目的元信息。
    /// </summary>
    public static List<IconEntryInfo> ParseEntries(byte[] icoData)
    {
        var entries = new List<IconEntryInfo>();
        if (icoData.Length < 6) return entries;

        var count = BitConverter.ToUInt16(icoData, 4);
        for (int i = 0; i < count; i++)
        {
            var entryOffset = 6 + i * 16;
            if (entryOffset + 16 > icoData.Length) break;

            int w = icoData[entryOffset] == 0 ? 256 : icoData[entryOffset];
            int h = icoData[entryOffset + 1] == 0 ? 256 : icoData[entryOffset + 1];
            int bitCount = BitConverter.ToUInt16(icoData, entryOffset + 6);
            int dataSize = BitConverter.ToInt32(icoData, entryOffset + 8);

            entries.Add(new IconEntryInfo(w, h, bitCount, dataSize));
        }

        return entries;
    }

    /// <summary>
    /// 获取 ICO 中最大尺寸条目的索引。
    /// </summary>
    public static int GetLargestEntryIndex(byte[] icoData)
    {
        var entries = ParseEntries(icoData);
        if (entries.Count == 0) return -1;

        int maxIndex = 0;
        int maxPixels = 0;
        for (int i = 0; i < entries.Count; i++)
        {
            int pixels = entries[i].Width * entries[i].Height;
            if (pixels > maxPixels)
            {
                maxPixels = pixels;
                maxIndex = i;
            }
        }
        return maxIndex;
    }

    /// <summary>
    /// 描述 ICO 文件中包含的尺寸列表。
    /// </summary>
    public static string DescribeSizes(byte[] icoData)
    {
        var entries = ParseEntries(icoData);
        if (entries.Count == 0) return "?";

        var sizes = entries.Select(e => e.Width == e.Height ? $"{e.Width}" : $"{e.Width}x{e.Height}");
        return $"{entries.Count} 张 ({string.Join("/", sizes)})";
    }

    /// <summary>
    /// 获取 ICO 中最大尺寸。
    /// </summary>
    public static int GetMaxSize(byte[] icoData)
    {
        var entries = ParseEntries(icoData);
        return entries.Count > 0 ? entries.Max(e => Math.Max(e.Width, e.Height)) : 0;
    }

    /// <summary>
    /// 从 ICO 文件字节中提取指定条目的原始图像数据。
    /// 返回的数据可能是 PNG（以 0x89504E47 开头）或 BMP DIB 数据。
    /// </summary>
    public static byte[]? ExtractEntryImageData(byte[] icoData, int entryIndex)
    {
        if (icoData.Length < 6) return null;

        var count = BitConverter.ToUInt16(icoData, 4);
        if (entryIndex < 0 || entryIndex >= count) return null;

        int dirEntryOffset = 6 + entryIndex * 16;
        if (dirEntryOffset + 16 > icoData.Length) return null;

        int dataSize = BitConverter.ToInt32(icoData, dirEntryOffset + 8);
        int dataOffset = BitConverter.ToInt32(icoData, dirEntryOffset + 12);

        if (dataOffset + dataSize > icoData.Length || dataSize <= 0) return null;

        var data = new byte[dataSize];
        Array.Copy(icoData, dataOffset, data, 0, dataSize);
        return data;
    }

    /// <summary>
    /// 从 ICO 数据中提取最大尺寸的图像数据（PNG 或 BMP DIB）。
    /// </summary>
    public static byte[]? ExtractLargestImageData(byte[] icoData)
    {
        int idx = GetLargestEntryIndex(icoData);
        return idx >= 0 ? ExtractEntryImageData(icoData, idx) : null;
    }

    /// <summary>
    /// 判断字节数据是否为 PNG 格式。
    /// </summary>
    public static bool IsPng(byte[] data)
    {
        return data.Length >= 8
            && data[0] == 0x89 && data[1] == 0x50
            && data[2] == 0x4E && data[3] == 0x47;
    }
}

File: tools/rust/core/src/ico_helper.rs
use crate::models::IconEntryInfo;

/// ICO 文件格式解析与操作工具。

/// 解析 ICO 文件字节，返回所有图像条目的元信息。
pub fn parse_entries(ico_data: &[u8]) -> Vec<IconEntryInfo> {
    let mut entries = Vec::new();
    if ico_data.len() < 6 {
        return entries;
    }

    let count = u16::from_le_bytes([ico_data[4], ico_data[5]]) as usize;
    for i in 0..count {
        let entry_offset = 6 + i * 16;
        if entry_offset + 16 > ico_data.len() {
            break;
        }

        let w = if ico_data[entry_offset] == 0 { 256 } else { ico_data[entry_offset] as u32 };
        let h = if ico_data[entry_offset + 1] == 0 { 256 } else { ico_data[entry_offset + 1] as u32 };
        let bit_count = u16::from_le_bytes([
            ico_data[entry_offset + 6],
            ico_data[entry_offset + 7],
        ]) as u32;
        let data_size = u32::from_le_bytes([
            ico_data[entry_offset + 8],
            ico_data[entry_offset + 9],
            ico_data[entry_offset + 10],
            ico_data[entry_offset + 11],
        ]);

        entries.push(IconEntryInfo {
            width: w,
            height: h,
            bit_count,
            data_size,
        });
    }

    entries
}

/// 获取 ICO 中最大尺寸条目的索引。
pub fn get_largest_entry_index(ico_data: &[u8]) -> Option<usize> {
    let entries = parse_entries(ico_data);
    if entries.is_empty() {
        return None;
    }

    let mut max_index = 0;
    let mut max_pixels: u64 = 0;
    for (i, entry) in entries.iter().enumerate() {
        let pixels = entry.width as u64 * entry.height as u64;
        if pixels > max_pixels {
            max_pixels = pixels;
            max_index = i;
        }
    }
    Some(max_index)
}

/// 描述 ICO 文件中包含的尺寸列表。
pub fn describe_sizes(ico_data: &[u8]) -> String {
    let entries = parse_entries(ico_data);
    if entries.is_empty() {
        return "?".to_string();
    }

    let sizes: Vec<String> = entries
        .iter()
        .map(|e| {
            if e.width == e.height {
                format!("{}", e.width)
            } else {
                format!("{}x{}", e.width, e.height)
            }
        })
        .collect();

    format!("{} 张 ({})", entries.len(), sizes.join("/"))
}

/// 获取 ICO 中最大尺寸。
pub fn get_max_size(ico_data: &[u8]) -> u32 {
    let entries = parse_entries(ico_data);
    entries
        .iter()
        .map(|e| e.width.max(e.height))
        .max()
        .unwrap_or(0)
}

/// 从 ICO 文件字节中提取指定条目的原始图像数据。
/// 返回的数据可能是 PNG（以 0x89504E47 开头）或 BMP DIB 数据。
pub fn extract_entry_image_data(ico_data: &[u8], entry_index: usize) -> Option<Vec<u8>> {
    if ico_data.len() < 6 {
        return None;
    }

    let count = u16::from_le_bytes([ico_data[4], ico_data[5]]) as usize;
    if entry_index >= count {
        return None;
    }

    let dir_entry_offset = 6 + entry_index * 16;
    if dir_entry_offset + 16 > ico_data.len() {
        return None;
    }

    let data_size = u32::from_le_bytes([
        ico_data[dir_entry_offset + 8],
        ico_data[dir_entry_offset + 9],
        ico_data[dir_entry_offset + 10],
        ico_data[dir_entry_offset + 11],
    ]) as usize;
    let data_offset = u32::from_le_bytes([
        ico_data[dir_entry_offset + 12],
        ico_data[dir_entry_offset + 13],
        ico_data[dir_entry_offset + 14],
        ico_data[dir_entry_offset + 15],
    ]) as usize;

    if data_offset + data_size > ico_data.len() || data_size == 0 {
        return None;
    }

    Some(ico_data[data_offset..data_offset + data_size].to_vec())
}

/// 从 ICO 数据中提取最大尺寸的图像数据（PNG 或 BMP DIB）。
pub fn extract_largest_image_data(ico_data: &[u8]) -> Option<Vec<u8>> {
    let idx = get_largest_entry_index(ico_data)?;
    extract_entry_image_data(ico_data, idx)
}

/// 判断字节数据是否为 PNG 格式。
pub fn is_png(data: &[u8]) -> bool {
    data.len() >= 8 && data[0] == 0x89 && data[1] == 0x50 && data[2] == 0x4E && data[3] == 0x47
}
